scripts/importer.py

```diff
@@ -27,51 +27,171 @@
 
 
 import argparse
+import sys
+
+browser_default_input_format = {
+    'chromium': 'netscape',
+    'ie': 'netscape',
+    'firefox': 'netscape',
+    'seamonkey': 'netscape',
+    'palemoon': 'netscape'
+}
 
 
 def main():
     args = get_args()
-    if args.browser in ['chromium', 'firefox', 'ie']:
-        import_netscape_bookmarks(args.bookmarks, args.bookmark_format)
+    bookmark_types = []
+    output_format = None
+    input_format = args.input_format
+    if args.import_search:
+        bookmark_types = ['search']
+        if args.oldconfig:
+            output_format = 'oldsearch'
+        else:
+            output_format = 'search'
+    else:
+        if args.bookmark_output:
+            output_format = 'bookmark'
+        elif args.quickmark_output:
+            output_format = 'quickmark'
+        if args.import_bookmarks:
+            bookmark_types.append('bookmark')
+        if args.import_keywords:
+            bookmark_types.append('keyword')
+    if not bookmark_types:
+        bookmark_types = ['bookmark', 'keyword']
+    if not output_format:
+        output_format = 'quickmark'
+    if not input_format:
+        if not args.browser:
+            sys.exit("Must specify either browser or input format")
+        input_format = browser_default_input_format[args.browser]
+
+    import_function = {'netscape': import_netscape_bookmarks}
+    import_function[input_format](args.bookmarks, bookmark_types,
+                                  output_format)
 
 
 def get_args():
     """Get the argparse parser."""
     parser = argparse.ArgumentParser(
         epilog="To import bookmarks from Chromium, Firefox or IE, "
-               "export them to HTML in your browsers bookmark manager. "
-               "By default, this script will output in a quickmarks format.")
-    parser.add_argument('browser', help="Which browser? (chromium, firefox)",
-                        choices=['chromium', 'firefox', 'ie'],
-                        metavar='browser')
-    parser.add_argument('-b', help="Output in bookmark format.",
-                        dest='bookmark_format', action='store_true',
-                        default=False, required=False)
+        "export them to HTML in your browsers bookmark manager. ")
+    parser.add_argument(
+        'browser',
+        help="Which browser? {%(choices)s}",
+        choices=browser_default_input_format.keys(),
+        nargs='?',
+        metavar='browser')
+    parser.add_argument(
+        '-i',
+        '--input-format',
+        help='Which input format? (overrides browser default)',
+        choices=set(browser_default_input_format.values()),
```

Comment:
Since this only has one default value  `{netscape}`, can't this be a default?

Reply:
I have it set like that because it's useful when other formats are available. I'm not particularly sure if it would be appropriate to actually merge it into this PR, but my Mozilla support patch

```diff --git a/scripts/importer.py b/scripts/importer.py
index 5886efa5a..9fd701be7 100755
--- a/scripts/importer.py
+++ b/scripts/importer.py
@@ -27,14 +27,16 @@ Currently only importing bookmarks from Netscape Bookmark files is supported.
 
 
 import argparse
+import sqlite3
 import sys
+import os
 
 browser_default_input_format = {
     'chromium': 'netscape',
     'ie': 'netscape',
-    'firefox': 'netscape',
-    'seamonkey': 'netscape',
-    'palemoon': 'netscape'
+    'firefox': 'mozilla',
+    'seamonkey': 'mozilla',
+    'palemoon': 'mozilla'
 }
 
 
@@ -67,7 +69,10 @@ def main():
             sys.exit("Must specify either browser or input format")
         input_format = browser_default_input_format[args.browser]
 
-    import_function = {'netscape': import_netscape_bookmarks}
+    import_function = {
+        'netscape': import_netscape_bookmarks,
+        'mozilla': import_moz_places
+    }
     import_function[input_format](args.bookmarks, bookmark_types,
                                   output_format)
 
@@ -200,5 +205,62 @@ def import_netscape_bookmarks(bookmarks_file, bookmark_types, output_format):
         print(bookmark)
 
 
+def import_moz_places(profile, bookmark_types, output_format):
+    """Import bookmarks from a Mozilla profile's places.sqlite database."""
+    place_query = {
+        'bookmark':
+        ("SELECT DISTINCT moz_bookmarks.title,moz_places.url "
+         "FROM moz_bookmarks,moz_places "
+         "WHERE moz_places.id=moz_bookmarks.fk "
+         "AND moz_places.id NOT IN (SELECT place_id FROM moz_keywords) "
+         "AND moz_places.url NOT LIKE 'place:%';"
+         ),  # Bookmarks with no keywords assigned
+        'keyword':
+        ("SELECT moz_keywords.keyword,moz_places.url "
+         "FROM moz_keywords,moz_places,moz_bookmarks "
+         "WHERE moz_places.id=moz_bookmarks.fk "
+         "AND moz_places.id=moz_keywords.place_id "
+         "AND moz_places.url NOT LIKE '%!%s%' ESCAPE '!';"
+         ),  # Bookmarks with keywords assigned but no %s substitution
+        'search':
+        ("SELECT moz_keywords.keyword, "
+         "    moz_bookmarks.title, "
+         "    search_conv(moz_places.url) AS url "
+         "FROM moz_keywords,moz_places,moz_bookmarks "
+         "WHERE moz_places.id=moz_bookmarks.fk "
+         "AND moz_places.id=moz_keywords.place_id "
+         "AND moz_places.url LIKE '%!%s%' ESCAPE '!';"
+         )  # bookmarks with keyword and %s substitution
+    }
+    out_template = {
+        'bookmark': {
+            'bookmark': '{url} {title}',
+            'keyword': '{url} {keyword}'
+        },
+        'quickmark': {
+            'bookmark': '{title} {url}',
+            'keyword': '{keyword} {url}'
+        },
+        'oldsearch': {
+            'search': '{keyword} {url} #{title}'
+        },
+        'search': {
+            'search': "c.url.searchengines['{keyword}'] = '{url}' #{title}"
+        }
+    }
+
+    def search_conv(url):
+        return search_escape(url).replace('%s', '{}')
+
+    places = sqlite3.connect(os.path.join(profile, "places.sqlite"))
+    places.create_function('search_conv', 1, search_conv)
+    places.row_factory = sqlite3.Row
+    c = places.cursor()
+    for typ in bookmark_types:
+        c.execute(place_query[typ])
+        for row in c:
+            print(out_template[output_format][typ].format(**row))
+
+
 if __name__ == '__main__':
     main()
```
might make the intention more obvious.

Reply:
I think that netscape should still be the default (since it's more common and cross platform), but we should definetly keep the flag there for whene we add more import formats. I think that we should probably merge this as is and follow up with another PR for the second bit, it's good to keep PR's as small as possible :)

Reply:
I would definitely agree with that. It's set to the default in `main()` logic, as keeping the old command line behavior means overriding unset arguments based on other arguments. Would it be better to explicitly set the default to `None` when this is needed?

Reply:
My annoyance with it is that right now, the only supported option is `./importer -i netscape <bookmark file>`, or a bunch of `-b` options linking to netscape, so I really think it should just be truly default. IE: I should be able to run `./importer <bookmark file>` and have it just work (since there's only one supported format anyway.

Right now when you run `./importer <bookmark file>` it complains saying `Must specify either browser or input format`.

Reply:
Sorry about that; weird filesystem error meant the comment went through, but the commit did not.